runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/ArtifactSource.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.fnexecution.artifact;
+
+import io.grpc.stub.StreamObserver;
+import java.io.IOException;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.ArtifactChunk;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.Manifest;
+
+/**
+ * Makes artifacts available to an ArtifactRetrievalService by
+ * encapsulating runner-specific resources.
+ */
+public interface ArtifactSource {
+
+  /**
+   * Get the artifact manifest available from this source.
+   */
+  Manifest getManifest() throws IOException;
+
+  /**
+   * Get an artifact by its name.
+   */
+  void getArtifact(String name, StreamObserver<ArtifactChunk> responseObserver);
```

Comment:
Do we want to make clients interact directly with StreamObserver? This may be the most reasonable approach, but it also increases the burden on implementors due to gRPC semantics. For example, clients may need to wrap code to ensure that `onCompleted` or `onError` is always called to make sure RPCs complete in a timely fashion. If this is necessary, please add a doc comment.

Reply:
Maybe it's possible to implement a `getInputStream()` method? I'm not sure it's worth it though, since this class is directly used to implement ArtifactRetrievalService, which speaks in the same StreamObserver, right?

Reply:
Alternatively this could return an `Iterable<ArtifactChunk>`, where the chunks are constructed from the stream by the service implementation

Reply:
Ideally we'd be able to write a utility that takes the `InputStream` and produces the stream of chunks of the appropriate type, and the actual service implementation is responsible for pushing the bytes that the source makes available, so we don't have to write multiple sources that know how to get input streams from names but don't know how to turn those streams into chunks, or streams of chunks.